### DISCUSSION TITLE
Is preferredRegions just a priority list?

### DISCUSSION THREAD

**Discussion by kasperaamodt** [vercel/community#4285] [Help] [answered]
### Summary

Hi,

I'm trying to make an edge runtime function that reads the region it is executing in, using the x-vercel-id header. While doing this, I found an unexpected behaviour. I set my preferred regions to ['fra1', 'cle1']. Then I used a VPN in San Fransisco, and the header read sfo1::fra1. This indicated that the closest region is sfo1, which is not in my preferred regions array, so it just selected the first one in the array instead. I thought that the config would select the closest region in the array, not the first one if it's not a direct hit. Am I doing something wrong, or is this how it actually works? If it is, it seems like a priority list.

### Steps to Reproduce

Described in the summary.

**Comment by amyegan** [comment]
With Edge Functions, visitors automatically use the region that is geographically closest to them. You can set the `config.regions` key or `preferredRegions` in your Edge Function for more control over where it should execute, but note that it will automatically failover to other regions to maintain uptime in the event of an outage.

There are also over 100 points of presence (POPs) in front of our compute-capable regions. Traffic to Vercel is routed to these POPs, and then from the POPs to the nearest region to get traffic to and from the Vercel Edge Network as quickly and efficiently as possible.

The following pages have much more information about our Edge Network:
- [Can I redirect users to a specific region?](https://vercel.com/docs/edge-network/frequently-asked-questions#can-i-redirect-users-to-a-specific-region)
- [Regional Edge Function invocation](https://vercel.com/docs/functions/edge-functions#regional-edge-function-invocation)
- [Regions and content delivery](https://vercel.com/docs/edge-network/regions#regions-and-content-delivery)

**Reply by kasperaamodt** [reply]
Ok, I am using next to get that out of the way.

Can you be more specific in regards to my question? In the event that I have a list of `preferredRegions`, does it use the closest one in that list to the user, or the first one, if the actually closest is not in the list?

**Reply by amyegan** [reply]
By default it chooses the nearest region, but it switches to the order provided when you specify regions.

I also just noticed a typo in our conversation. It should be [`preferredRegion`](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#preferredregion) rather than `preferredRegions`

**Reply by kasperaamodt** [reply]
Ok, and it is not possible to get the best of both worlds here? Nearest region in the list?

**Comment by matannahmani** [comment]
@amyegan @kasperaamodt are there any updates about this one? I've also noticed that it works similarly to a priority list instead of actually routing to the nearest region from the list.

**Reply by amyegan** [reply]
Hi @matannahmani! There hasn't been any recent change to the `preferredRegion` behavior. By default edge functions use the nearest available region. If you specify a list of preferred regions, then it will instead be deployed to the specified regions and prioritize them based on the order of the list.

---

For example, I deployed an edge function with `export const preferredRegion = ['dub1', 'fra1'];`. The function is deployed to both regions, but prefers `dub1`.

<img width="1262" alt="deployment summary showing edge function deployed to Dublin and Frankfurt" src="https://github.com/vercel/community/assets/8247105/ea4fa348-75c1-4945-a7d8-0949f2dd4d35">


<img width="333" alt="log showing edge function ran in Dublin" src="https://github.com/vercel/community/assets/8247105/d39f79db-c2cc-4b15-bbc2-6f83289c275a">

-----

When I change it to `export const preferredRegion = ['fra1', 'dub1'];`, then we see that preference reflected in the deployment.

<img width="1248" alt="deployment summary showing edge function deployed to Frankfurt and Dublin" src="https://github.com/vercel/community/assets/8247105/2bb83722-8e66-4823-b64b-eead34e12868">

<img width="324" alt="log showing edge function ran in Frankfurt" src="https://github.com/vercel/community/assets/8247105/e2e6dd7b-a0db-4978-8224-6eebfcda7e82">

**Reply by matannahmani** [reply]
Hey @amyegan I feel like this is a bit counter intuitive, I believe that in the docs it doesn't mention the priority list or maybe I missed that part, but I would expect the edge function work as nearest location to the user but only the locations from the list.
Is there any chance of this being implemented in the future? Otherwise the current implementation is quite useless if I'm frankly honest.

Ideally we would use this preferred Region to match our database region but with the current implementation it's impossible if I understand correctly.
Example would be:
Database regions: fra1 iad1 hnd1
Edge function regions: fra1 iad1 hnd1
Incoming Request (Seoul) > edge function (hnd1) > hnd1 database

**Comment by smaeda-ks** [comment, marked as answer]
@matannahmani @kasperaamodt We're rolling out a fix this week to support specifying multiple regions for edge functions, so they work as you expect.

**Reply by smaeda-ks** [reply]
@matannahmani @kasperaamodt The fix has been rolled out. This should work now

**Comment by qtnx** [comment]
This configuration does not work as expected when used with ENV variables.​

I set the variable to an array of ['sin1', 'hnd1'] (with parsing from string 'sin1,hnd1' of course) but it still routes to hkg1.​

Hard coding works, but this should be fixed or changed as OpenAI is going to block requests from hkg1.

**Reply by leerob** [reply]
The values need to be statically analyzable, which is why you can't use an env var here.
